src/value.h

```diff
@@ -85,159 +93,185 @@ class RangeType {
+	const char* c_str() const noexcept { return this->str_ptr->first.c_str(); }
+	const std::string& toString() const noexcept { return this->str_ptr->first; }
+	size_t size() const noexcept { return this->str_ptr->first.size(); }
 
-class str_utf8_wrapper : public std::string
-{
-public:
-	str_utf8_wrapper() : std::string(), cached_len(-1) { }
-	str_utf8_wrapper( const std::string& s ) : std::string( s ), cached_len(-1) { }
-	str_utf8_wrapper( size_t n, char c ) : std::string(n, c), cached_len(-1) { }
-	~str_utf8_wrapper() {}
-	
 	glong get_utf8_strlen() const {
-		if (cached_len < 0) {
-			cached_len = g_utf8_strlen(this->c_str(), this->size());
+		if (str_ptr->second < 0) {
+			str_ptr->second = g_utf8_strlen(this->str_ptr->first.c_str(), this->str_ptr->first.size());
 		}
-		return cached_len;
+		return str_ptr->second;
 	};
+
+
 private:
-	mutable glong cached_len;
+	str_ptr_t str_ptr;
 };
 
-
 class Value
 {
 public:
-	typedef std::vector<ValuePtr> VectorType;
-
-  enum class ValueType {
-    UNDEFINED,
-    BOOL,
-    NUMBER,
-    STRING,
-    VECTOR,
-    RANGE
-  };
-  static const Value undefined;
-
-  Value();
-  Value(bool v);
-  Value(int v);
-  Value(double v);
-  Value(const std::string &v);
-  Value(const char *v);
-  Value(const char v);
-  Value(const VectorType &v);
-  Value(const RangeType &v);
-  ~Value() {}
-
-  ValueType type() const;
-  bool isDefined() const;
-  bool isDefinedAs(const ValueType type) const;
-  bool isUndefined() const;
-
-  double toDouble() const;
-  bool getDouble(double &v) const;
-  bool getFiniteDouble(double &v) const;
-  bool toBool() const;
-  std::string toString() const;
-  std::string toString(const tostring_visitor *visitor) const;
-  std::string toEchoString() const;
-  std::string toEchoString(const tostring_visitor *visitor) const;
-  void toStream(std::ostringstream &stream) const;
-  void toStream(const tostream_visitor *visitor) const;
-  std::string chrString() const;
-  const VectorType &toVector() const;
-  bool getVec2(double &x, double &y, bool ignoreInfinite = false) const;
-  bool getVec3(double &x, double &y, double &z) const;
-  bool getVec3(double &x, double &y, double &z, double defaultval) const;
-  RangeType toRange() const;
+	typedef std::vector<Value> VectorType;
+	enum class ValueType {
+		UNDEFINED,
+		BOOL,
+		NUMBER,
+		STRING,
+		VECTOR,
+		RANGE
+	};
+	static const Value undefined;
+
+	class VectorPtr {
+	public:
+		VectorPtr() : ptr(make_shared<VectorType>()) {}
+		VectorPtr(double x, double y, double z);
+		VectorPtr(const VectorPtr &) = delete; // never copy, move instead
+		VectorPtr& operator=(const VectorPtr &) = delete; // never copy, move instead
+		VectorPtr(VectorPtr&&) = default; // default move constructor
+		VectorPtr& operator=(VectorPtr&&) = default; // default move assignment
+
+		// Copy explicitly only when necessary
+		// We can't use unique_ptr because of this :(
+		VectorPtr clone() const { VectorPtr c; c.ptr = this->ptr; return c; }
```

Comment:
This clones the shared_ptr, of course. To avoid silly mistakes, maybe this (and other pointer-clone methods) could be called ```clonePtr()``` or so? Then keep the real ```clone()``` for Value.
In practice, sometimes you may have code like this that does not immediately show that a pointer is being cloned:
```
vec = result.clone();
```

Reply:
See my reply below about the interface changes which should make this less of an issue.
This comment did cause me to think harder about when it's safe to move elements out of vectors and ultimately resulted in the new `EmbeddedVectorType` which has shown great benefits, so I do appreciate the input.